ddtrace/llmobs/_integrations/vllm.py

```diff
@@ -0,0 +1,165 @@
+from ddtrace.llmobs._constants import INPUT_TOKENS_METRIC_KEY
+from ddtrace.llmobs._constants import METADATA
+from ddtrace.llmobs._constants import METRICS
+from ddtrace.llmobs._constants import MODEL_NAME
+from ddtrace.llmobs._constants import MODEL_PROVIDER
+from ddtrace.llmobs._constants import OUTPUT_MESSAGES
+from ddtrace.llmobs._constants import OUTPUT_TOKENS_METRIC_KEY
+from ddtrace.llmobs._constants import OUTPUT_VALUE
+from ddtrace.llmobs._constants import SPAN_KIND
+from ddtrace.llmobs._constants import TIME_IN_MODEL_DECODE_METRIC_KEY
+from ddtrace.llmobs._constants import TIME_IN_MODEL_INFERENCE_METRIC_KEY
+from ddtrace.llmobs._constants import TIME_IN_MODEL_PREFILL_METRIC_KEY
+from ddtrace.llmobs._constants import TIME_IN_QUEUE_METRIC_KEY
+from ddtrace.llmobs._constants import TIME_TO_FIRST_TOKEN_METRIC_KEY
+from ddtrace.llmobs._constants import TOTAL_TOKENS_METRIC_KEY
+from ddtrace.llmobs._integrations.base import BaseLLMIntegration
+from ddtrace.llmobs.types import Message
+from ddtrace.llmobs.utils import Document
+from ddtrace.trace import Span
+
+
+class VLLMIntegration(BaseLLMIntegration):
+    """LLMObs integration for vLLM V1 library."""
+
+    _integration_name = "vllm"
+
+    _METADATA_FIELDS = {
+        "temperature",
+        "max_tokens",
+        "top_p",
+        "n",
+        "num_cached_tokens",
+        "embedding_dim",
+        "finish_reason",
+        "lora_name",
+    }
+
+    def _set_base_span_tags(self, span: Span, **kwargs: Any) -> None:
+        """Set base tags on vLLM spans."""
+        model_name = kwargs.get("model_name")
+        if model_name:
+            span._set_tag_str(TAG_MODEL, model_name)
+            span._set_tag_str(TAG_PROVIDER, PROVIDER_NAME)
+
+    def _build_metadata(self, data: RequestData) -> Dict[str, Any]:
+        """Extract metadata from request data."""
+        md: Dict[str, Any] = {}
+
+        for key in self._METADATA_FIELDS:
+            val = getattr(data, key, None)
+            if val is not None:
+                md[key] = val
+
+        return md
+
+    def _build_metrics(self, data: RequestData, latency_metrics: Optional[LatencyMetrics] = None) -> Dict[str, Any]:
+        """Build token and latency metrics from request data."""
+        it = int(data.input_tokens or 0)
+        ot = int(data.output_tokens or 0)
+        metrics: Dict[str, Any] = {
+            INPUT_TOKENS_METRIC_KEY: it,
+            OUTPUT_TOKENS_METRIC_KEY: ot,
+            TOTAL_TOKENS_METRIC_KEY: it + ot,
+        }
+
+        if latency_metrics:
+            metric_map = {
+                "time_to_first_token": TIME_TO_FIRST_TOKEN_METRIC_KEY,
+                "time_in_queue": TIME_IN_QUEUE_METRIC_KEY,
+                "time_in_model_prefill": TIME_IN_MODEL_PREFILL_METRIC_KEY,
+                "time_in_model_decode": TIME_IN_MODEL_DECODE_METRIC_KEY,
+                "time_in_model_inference": TIME_IN_MODEL_INFERENCE_METRIC_KEY,
+            }
```

Comment:
This looks like the same metric map as utils.py line 50, all the more reason to make it a constant?

Reply:
Althought I guess they serve different purposes... hmm up to you if you want to reuse, but at least make this one a class constant if you don't want them to be the same.